pkg/installer/upgrade.go

```diff
@@ -29,39 +29,40 @@ func Upgrade(uninstallConfig *apiv1.KubectlStorageOSConfig, installConfig *apiv1
 		return err
 	}
 
-	// discover existing secret username and password for upgrade. Here we use the (un)installer
-	// as it contains the manifests to be uninstalled, and the installConfig so we can set existing
-	// secret username and password in the secret manifest to be installed
-	err = uninstaller.storeExistingStorageOSSecretData(installConfig)
+	// uninstall existing storageos operator and cluster
+	err = uninstaller.Uninstall(uninstallConfig, versionToUninstall)
+	if err != nil {
+		return err
+	}
+
+	// discover uninstalled secret username and password for upgrade. Here we use (1) the (un)installer
+	// as it contains the on-disk FS of the uninstalled secrets and (2) the installConfig so we can
+	// set secret username and password in the secret manifest to be installed
+	err = uninstaller.copyStorageOSSecretData(installConfig)
 	if err != nil {
 		return err
 	}
 
 	// if the version being uninstalled during upgrade is that of the 'old' operator (pre v2.5) existing
-	// CSI secrets and 'fast' storage class must be stored so that they can be recreated after upgrade.
-	storeData, err := pluginversion.VersionIsLessThanOrEqual(versionToUninstall, pluginversion.ClusterOperatorLastVersion())
+	// CSI secrets and 'fast' storage class must be recreated
+	storageosV1, err := pluginversion.VersionIsLessThanOrEqual(versionToUninstall, pluginversion.ClusterOperatorLastVersion())
 	if err != nil {
 		return err
 	}
-	var csiSecrets []*corev1.Secret
-	var storageClass *kstoragev1.StorageClass
-	if storeData {
-		// discover and store the old (pre v2.5) CSI secrets in kube-system
-		csiSecrets, err = uninstaller.storeExistingCSISecrets()
+
+	if storageosV1 {
+		// recreate storage class from uninstalled version, manifest saved to disk
+		err = uninstaller.recreateStorageClass()
```

Comment:
Recreating storage class will change UID of it, which is stored in PVC. If api manager founds different SC UID, it stops volume sync :( I'm not sure what this change would cause.

Reply:
Hmm.. How about instead of recreating, we add a finalizer to the storage class before uninstall so that it is not deleted?

Reply:
We have to test it first. Finalizer increases the deletion time and should end in timeout.